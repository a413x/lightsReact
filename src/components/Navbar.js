import React, { Component } from "react";
import { Link } from 'react-router-dom';

import '../styles/nav.css';

class Navbar extends Component{
  constructor(props){
    super(props);
    this.state = {
      inputsVal: this.props.times
    };
  }
  inputChange(e,ind){
    const val = parseInt(e.target.value);
    if(!val || val > 99 || val < 0) return;
    const nextTimes = [...this.state.inputsVal];
    nextTimes[ind] = val;
    this.setState({inputsVal:nextTimes});
  }
  onSetClick(){
    this.props.onSetClick([...this.state.inputsVal]);
  }
  render(){
    const colors = this.props.colors;
    return <div className = 'nav'>
      {colors.map((el,ind)=>
        <div key={el.id}>
          <Link to={'/'+el.color} onClick={()=>this.props.onLinkClick(ind)}>
            {el.color.charAt(0).toUpperCase() + el.color.slice(1)}
          </Link>
          <input type='number' min = '0' max = '100'
            value = {this.state.inputsVal[ind]} onChange = {(e)=>{this.inputChange(e,ind)}}/>
          <button onClick = {()=>this.onSetClick()}>Set</button>
        </div>
      )}
    </div>
  }
}

export default Navbar;
