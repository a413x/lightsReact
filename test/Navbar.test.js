import React from 'react';
import { shallow, mount } from 'enzyme';
import { createMemoryHistory } from "history";
import { Router } from "react-router";

import Navbar from '../src/components/Navbar.js';

describe('Tests for Navbar component',()=>{
  const mockOnSetClick = jest.fn();
  const mockOnLinkClick = jest.fn();

  const props = {
    colors : [
      {id:0,color:'red'},
      {id:1,color:'yellow'},
      {id:2,color:'green'},
    ],
    times : [5, 4, 7],
    onSetClick: mockOnSetClick,
    onLinkClick: mockOnLinkClick
  };

  const initialState = {
    inputsVal: props.times
  };

  const container = shallow(<Navbar {...props} />);

  describe('Navbar initial',()=>{
    it('renders properly',()=>{
      expect(container).toMatchSnapshot();
    });
    it('initialize with initial state',()=>{
      expect(container.state()).toEqual(initialState);
    });
  });

  describe('Navbar links',()=>{
    let history = createMemoryHistory();
    const routerContainer = mount(
      <Router history = {history}>
        <Navbar {...props} />
      </Router>,
    );
    for(let i = 0; i < props.colors.length; i++){
      let color = '/' + props.colors[i].color;
      it('clicked on link ' + color,()=>{
        routerContainer.find('Link').filter({to:color}).simulate('click',{ button: 0 });
        expect(mockOnLinkClick).toHaveBeenCalledWith(i);
        expect(history.location.pathname).toBe(color);
      });
    }
    afterAll(()=>{
      routerContainer.unmount();
    })
  });

  describe('Navbar inputs',()=>{
    it('simulate change on a first input with right value',()=>{
      container.find('input').first().simulate('change',{
        target:{value:'12'}
      });
      expect(container.state().inputsVal[0]).toBe(12);
    });
    it('simulate change on a last input with wrong value',()=>{
      container.find('input').last().simulate('change',{
        target:{value:'abc'}
      });
      expect(container.state().inputsVal[2]).toBe(props.times[2]);
    });
  });

  describe('Navbar buttons',()=>{
    for(let i = 0; i < props.colors.length; i++){
      it('onSetClick called when clicked on button next to ' + props.colors[i].color,()=>{
        container.find('button').at(i).simulate('click');
        expect(mockOnSetClick).toHaveBeenCalledWith(container.state().inputsVal);
      });
    }
  });

  afterAll(()=>{
    container.setState(initialState);
  });
});
