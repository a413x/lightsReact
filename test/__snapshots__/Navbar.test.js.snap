// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Navbar component Navbar initial renders properly 1`] = `
<div
  className="nav"
>
  <div
    key="0"
  >
    <Link
      onClick={[Function]}
      to="/red"
    >
      Red
    </Link>
    <input
      max="100"
      min="0"
      onChange={[Function]}
      type="number"
      value={5}
    />
    <button
      onClick={[Function]}
    >
      Set
    </button>
  </div>
  <div
    key="1"
  >
    <Link
      onClick={[Function]}
      to="/yellow"
    >
      Yellow
    </Link>
    <input
      max="100"
      min="0"
      onChange={[Function]}
      type="number"
      value={4}
    />
    <button
      onClick={[Function]}
    >
      Set
    </button>
  </div>
  <div
    key="2"
  >
    <Link
      onClick={[Function]}
      to="/green"
    >
      Green
    </Link>
    <input
      max="100"
      min="0"
      onChange={[Function]}
      type="number"
      value={7}
    />
    <button
      onClick={[Function]}
    >
      Set
    </button>
  </div>
</div>
`;
